src/compiler/WebSharper.FSharp.Service/Program.fs

```diff
@@ -81,83 +139,96 @@ let startListening() =
+                        override _.Error s =
+                            let sendErr = sprintf "e: %s" |> send
+                            let asyncValue = 
+                                s
+                                |> sendErr
+                            Async.RunSynchronously(asyncValue, cancellationToken = token)
+                    }
+
+                let processCompileMessage (projectOption: string option) wsConfig warnSettings args = async {
+                    nLogger.Debug(sprintf "Compiling %s" projectOption.Value)
+                    let compilationResultForDebugOrRelease() =
 #if DEBUG
-                                Compile wsConfig warnSettings logger checkerFactory tryGetMetadata
+                        Compile wsConfig warnSettings logger checkerFactory tryGetMetadata
 #else
-                                try Compile wsConfig warnSettings logger checkerFactory tryGetMetadata
-                                with 
-                                | ArgumentError msg -> 
-                                    PrintGlobalError logger (msg + " - args: " + (deserializedMessage.args |> String.concat " "))
-                                    1
-                                | e -> 
-                                    PrintGlobalError logger (sprintf "Global error: %A" e)
-                                    1
+                        try Compile wsConfig warnSettings logger checkerFactory tryGetMetadata
+                        with 
+                        | ArgumentError msg -> 
+                            PrintGlobalError logger (msg + " - args: " + (args |> String.concat " "))
+                            1
+                        | e -> 
+                            PrintGlobalError logger (sprintf "Global error: %A" e)
+                            1
 #endif
-                        let returnValue = compilationResultForDebugOrRelease()
-                        do! sendFinished returnValue
-                    | None ->
-                        ()
+                    let returnValue = compilationResultForDebugOrRelease()
+                    match projectOption with
+                    | Some project ->
+                        let projectTimestamp = { Path = project; Timestamp = File.GetLastWriteTime project }
+                        let referenceTimestamps = 
+                            wsConfig.References
+                            |> Array.map (fun x -> { Path = x; Timestamp = File.GetLastWriteTime x })
+                            |> Array.toList
+                        argsDict.[project] <- { Timestamps = projectTimestamp :: referenceTimestamps; Args = args }
+                    | None -> ()
+                    do! sendFinished returnValue
+                    }
+                match deserializedMessage with
+                | Exit -> 
+                    exiting <- true
+                | PostCompile project ->
+                    if argsDict.ContainsKey project |> not then
```

Comment:
Here too I would prefer TryGetValue instead of ContainsKey then lookup

you can match it nicely in F#:
`match d.TryGetValue(x) with | res, true -> ... | _ -> ...`